File: src/test/java/org/crafter_unit_tests/ChunkArrayTest.java
/*
 Crafter - A blocky game (engine) written in Java with LWJGL.
 Copyright (C) 2023  jordan4ibanez

 This program is free software: you can redistribute it and/or modify
 it under the terms of the GNU General Public License as published by
 the Free Software Foundation, either version 3 of the License, or
 (at your option) any later version.

 This program is distributed in the hope that it will be useful,
 but WITHOUT ANY WARRANTY; without even the implied warranty of
 MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 GNU General Public License for more details.

 You should have received a copy of the GNU General Public License
 along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */
package org.crafter_unit_tests;

import org.crafter.engine.world.chunk.Chunk;
import org.joml.Vector3ic;
import org.junit.jupiter.api.Test;

import static org.crafter.engine.utility.JOMLUtils.printVec;
import static org.junit.jupiter.api.Assertions.assertEquals;

public class ChunkArrayTest {

    @Test
    public void testChunkArrayIndexing() {

        for (int i = 0; i < Chunk.getArraySize(); i++) {
            Vector3ic position = Chunk.indexToPosition(i);
//            printVec(position);
            int index = Chunk.positionToIndex(position);
            assertEquals(index, i);
            Vector3ic doubleCheck = Chunk.indexToPosition(index);
            assertEquals(doubleCheck, position);
        }
    }

    @Test
    public void testConversions() {
        Chunk testChunk = new Chunk(0,0);

        // Basic test
        for (int i = 0; i < Chunk.getArraySize(); i++) {
            testChunk.setBlockData(i, i);
            int gottenBlockData = testChunk.getBlockData(i);

            assertEquals(i, gottenBlockData);

            Vector3ic position = Chunk.indexToPosition(i);
            final int newIndex = Chunk.positionToIndex(position);
            gottenBlockData = testChunk.getBlockData(position);

            assertEquals(newIndex, gottenBlockData);
            assertEquals(i, newIndex);
            assertEquals(i, gottenBlockData);
        }

        // Now test data stream
        final int[] workerChunkData = testChunk.getData();
        for (int i = 0; i < Chunk.getArraySize(); i++) {
            workerChunkData[i] = i + 10;
        }
        testChunk.setData(workerChunkData);
        for (int i = 0; i < Chunk.getArraySize(); i++) {
            final int gottenBlockData = testChunk.getBlockData(i);
            assertEquals(i + 10, gottenBlockData);
        }

        // Now test it again
        for (int i = 0; i < Chunk.getArraySize(); i++) {
            testChunk.setBlockData(i, i);
        }
        final int[] doubleCheck = testChunk.getData();
        for (int i = 0; i < Chunk.getArraySize(); i++) {
            assertEquals(doubleCheck[i], testChunk.getBlockData(i));
        }
    }
}

File: src/test/kotlin/org/crafter_unit_tests/ChunkArrayTest.kt
package org.crafter_unit_tests

import org.crafter.engine.world.chunk.Chunk
import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.Test

class ChunkArrayTest {
    @Test
    fun testChunkArrayIndexing() {
        val testChunk = Chunk(0, 0)
        for (i in 0 until testChunk.arraySize) {
            val position = testChunk.indexToPosition(i)
            val index = testChunk.positionToIndex(position)
            Assertions.assertEquals(index, i)
            val doubleCheck = testChunk.indexToPosition(index)
            Assertions.assertEquals(doubleCheck, position)
        }
    }

    @Test
    fun testConversions() {
        val testChunk = Chunk(0, 0)

        // Basic test
        for (i in 0 until testChunk.arraySize) {
            testChunk.setBlockData(i, i)
            var gottenBlockData = testChunk.getBlockData(i)
            Assertions.assertEquals(i, gottenBlockData)
            val position = testChunk.indexToPosition(i)
            val newIndex = testChunk.positionToIndex(position)
            gottenBlockData = testChunk.getBlockData(position)
            Assertions.assertEquals(newIndex, gottenBlockData)
            Assertions.assertEquals(i, newIndex)
            Assertions.assertEquals(i, gottenBlockData)
        }

        // Now test data stream
        val workerChunkData = testChunk.data
        for (i in 0 until testChunk.arraySize) {
            workerChunkData[i] = i + 10
        }
        testChunk.data = workerChunkData
        for (i in 0 until testChunk.arraySize) {
            val gottenBlockData = testChunk.getBlockData(i)
            Assertions.assertEquals(i + 10, gottenBlockData)
        }

        // Now test it again
        for (i in 0 until testChunk.arraySize) {
            testChunk.setBlockData(i, i)
        }
        val doubleCheck = testChunk.data
        for (i in 0 until testChunk.arraySize) {
            Assertions.assertEquals(doubleCheck[i], testChunk.getBlockData(i))
        }
    }
}
